flang/lib/Lower/Bridge.cpp

```diff
@@ -751,23 +751,27 @@ class FirConverter : public Fortran::lower::AbstractConverter {
         info.isStructured() ? builder->getIndexType() : info.loopVariableType;
     auto lowerValue = genFIRLoopIndex(info.lowerExpr, type);
     auto upperValue = genFIRLoopIndex(info.upperExpr, type);
-    info.stepValue = info.stepExpr.has_value()
-                         ? genFIRLoopIndex(*info.stepExpr, type)
-                         : info.isStructured()
-                               ? builder->create<mlir::ConstantIndexOp>(loc, 1)
-                               : builder->createIntegerConstant(
-                                     loc, info.loopVariableType, 1);
+    info.stepValue =
+        info.stepExpr.has_value() ? genFIRLoopIndex(*info.stepExpr, type)
+        : info.isStructured()
+            ? builder->create<mlir::ConstantIndexOp>(loc, 1)
+            : builder->createIntegerConstant(loc, info.loopVariableType, 1);
     assert(info.stepValue && "step value must be set");
     info.loopVariable = createTemp(loc, *info.loopVariableSym);
 
     // Structured loop - generate fir.loop.
     if (info.isStructured()) {
+      // Perform the default initial assignment of the DO variable.
+      auto lowerVal =
+          builder->createConvert(loc, info.loopVariableType, lowerValue);
```

Comment:
This looks like it may now be in both loop forms.  Could the unstructured instance of this assignment at old line 785 be moved up to before the specialization split to cover both cases?  (The convert is probably already accounted for.)

Reply:
Will do.